# Dự án: Phân tích ứng dụng Google Play Store
## Đề tài: Khảo sát, tiền xử lý và mô hình hóa dữ liệu Google Play Store
### Notebook 02: Tích hợp dữ liệu qua PostgreSQL (02_postgresql_pipeline.ipynb)

**Mục tiêu của Notebook:**
Notebook này đóng vai trò xây dựng **nền tảng dữ liệu tập trung** cho toàn bộ dự án AI. Thay vì đọc các file CSV thô phân tán và lặp đi lặp lại ở các bước sau, chúng ta sẽ đưa dữ liệu vào hệ quản trị cơ sở dữ liệu quan hệ **PostgreSQL**. Tại đây, các chỉ mục (INDEX), các quan hệ khóa, và đặc biệt là các **Summary Views (Aggregation)** sẽ được tạo dựng nhằm liên kết dữ liệu an toàn mà không làm nhân bản dòng bảng chính, đảm bảo dữ liệu đầu ra hoàn hảo cho mô hình học máy.

**Sơ đồ Pipeline dữ liệu:**
```
CSV Files (raw) ──> Import PostgreSQL ──> CREATE INDEX ──> Aggregation View (Reviews Summary) ──> Flat View (vw_googleplaystore_flat) ──> Python pd.read_sql()
```


## 0. Kiểm tra và tự động cài đặt thư viện thiếu (psycopg2 & SQLAlchemy)


In [1]:
import sys
import subprocess

# Tự động kiểm tra và cài đặt các thư viện trực quan hóa và database nếu thiếu
required_libs = ["pandas", "numpy", "matplotlib", "seaborn", "sqlalchemy", "psycopg2-binary"]
missing_libs = []
for lib in required_libs:
    lib_name = lib.split("-")[0] if "-" in lib else lib
    try:
        __import__(lib_name)
    except ImportError:
        missing_libs.append(lib)

if missing_libs:
    print(f"Đang tự động cài đặt các thư viện còn thiếu: {missing_libs}...")
    try:
        subprocess.check_call([sys.executable, "-m", "pip", "install", *missing_libs])
        print("Đã cài đặt xong các thư viện cần thiết!")
    except Exception as e:
        print("Không thể cài đặt tự động. Hãy tự chạy lệnh này trong terminal: !pip install " + " ".join(missing_libs))
else:
    print("Môi trường đã đầy đủ các thư viện cần thiết.")


Môi trường đã đầy đủ các thư viện cần thiết.


# II. Khởi tạo Database và Cấu trúc Bảng (DDL)
Các lệnh SQL DDL để khởi tạo cơ sở dữ liệu và bảng staging thô trong pgAdmin4 (được ghi lại dưới đây để phục vụ báo cáo):

### 1. Tạo Database
```sql
CREATE DATABASE googleplaystore_db;
```

### 2. Tạo cấu trúc Bảng Staging (`01_create_tables.sql`)
```sql
-- Tạo bảng chính googleplaystore
CREATE TABLE IF NOT EXISTS googleplaystore (
    App VARCHAR(255),
    Category VARCHAR(100),
    Rating VARCHAR(50),
    Reviews VARCHAR(50),
    Size VARCHAR(50),
    Installs VARCHAR(50),
    Type VARCHAR(50),
    Price VARCHAR(50),
    Content_Rating VARCHAR(100),
    Genres VARCHAR(255),
    Last_Updated VARCHAR(100),
    Current_Ver VARCHAR(255),
    Android_Ver VARCHAR(255)
);

-- Tạo bảng phụ chứa user reviews
CREATE TABLE IF NOT EXISTS googleplaystore_user_reviews (
    App VARCHAR(255),
    Translated_Review TEXT,
    Sentiment VARCHAR(50),
    Sentiment_Polarity VARCHAR(50),
    Sentiment_Subjectivity VARCHAR(50)
);
```


# III. Nạp dữ liệu vào cơ sở dữ liệu (Import Data)
Các câu lệnh nạp dữ liệu từ file CSV thô vào database bằng lệnh `COPY` / `\copy` trong pgAdmin4:

```sql
-- Nạp dữ liệu bảng googleplaystore
\copy googleplaystore FROM 'c:/HOC_HANH/Mon dang hoc/AI21301_DU_AN_1/Du_AN/data/raw/googleplaystore.csv' WITH CSV HEADER ENCODING 'utf8';

-- Nạp dữ liệu bảng googleplaystore_user_reviews
\copy googleplaystore_user_reviews FROM 'c:/HOC_HANH/Mon dang hoc/AI21301_DU_AN_1/Du_AN/data/raw/googleplaystore_user_reviews.csv' WITH CSV HEADER ENCODING 'utf8';
```


# IV. Kết nối Database từ Python (psycopg2 / SQLAlchemy)
Dưới đây là đoạn mã Python minh họa cách kết nối trực tiếp với PostgreSQL bằng `psycopg2` và `SQLAlchemy` theo tiêu chuẩn dự án.


In [1]:
import os
import sys
import pandas as pd
from sqlalchemy import create_engine

# Thông số kết nối PostgreSQL mẫu
PG_USER = "postgres"
PG_PASSWORD = "your_password"
PG_HOST = "localhost"
PG_PORT = "5432"
PG_DB = "googleplaystore_db"

print("--- THIẾT LẬP KẾT NỐI POSTGRESQL (MẪU) ---")
connection_uri = f"postgresql://{PG_USER}:{PG_PASSWORD}@{PG_HOST}:{PG_PORT}/{PG_DB}"
print(f"Connection URI: postgresql://{PG_USER}:***@{PG_HOST}:{PG_PORT}/{PG_DB}")

# Mẹo nhỏ: Notebook này tích hợp sẵn một Database Engine giả lập (SQLite in-memory)
# để chạy trực tiếp hiển thị kết quả nếu bạn không chạy PG cục bộ.
import sqlite3
sim_conn = sqlite3.connect(':memory:')
print("Đang khởi tạo Database giả lập SQLite để thực thi câu lệnh SQL trực tiếp...")


--- THIẾT LẬP KẾT NỐI POSTGRESQL (MẪU) ---
Connection URI: postgresql://postgres:***@localhost:5432/googleplaystore_db
Đang khởi tạo Database giả lập SQLite để thực thi câu lệnh SQL trực tiếp...


## 1. Kiểm tra Dữ liệu & Nạp dữ liệu thô vào database Staging


In [1]:
import os
import sys
import pandas as pd
import numpy as np

print("=== CHẨN ĐOÁN HỆ THỐNG & ĐƯỜNG DẪN ===")
print("Thư mục đang chạy (getcwd):", os.getcwd())

# Hàm tìm kiếm file thông minh nhắm mục tiêu (không blind scan toàn bộ ổ đĩa)
def find_data_file(file_name):
    # 1. Thử các đường dẫn tương đối trực tiếp (không phân biệt chữ hoa thường của folder)
    candidates = [
        "../data/raw", "../data/Raw", "../Data/Raw",
        "Du_AN/data/raw", "Du_AN/data/Raw", "Du_AN/Data/Raw", "du_an/data/raw",
        "data/raw", "data/Raw", "Data/Raw",
        "../../Du_AN/data/raw", "../../data/raw",
        "../../../Du_AN/data/raw", "../../../data/raw"
    ]
    
    for folder in candidates:
        for name in [file_name, file_name.lower(), file_name.upper()]:
            p = os.path.join(folder, name)
            if os.path.exists(p):
                print(f"Tìm thấy '{file_name}' tại đường dẫn tương đối: {p}")
                return os.path.abspath(p)

    # 2. Dành riêng cho WSL/Docker: Quét trực tiếp các mount point khả dĩ
    if os.name != 'nt':
        wsl_candidates = [
            "/mnt/c/HOC_HANH/Mon dang hoc/AI21301_DU_AN_1/Du_AN/data/raw",
            "/mnt/c/Hoc_Hanh/Mon dang hoc/AI21301_DU_AN_1/Du_AN/data/raw",
            "/mnt/c/Hoc Hanh/Mon dang hoc/AI21301_DU_AN_1/Du_AN/data/raw",
            "/mnt/d/HOC_HANH/Mon dang hoc/AI21301_DU_AN_1/Du_AN/data/raw",
            "/workspaces/AI21301_DU_AN_1/Du_AN/data/raw",
            "/workspace/AI21301_DU_AN_1/Du_AN/data/raw"
        ]
        for folder in wsl_candidates:
            for name in [file_name, file_name.lower(), file_name.upper()]:
                p = os.path.join(folder, name)
                if os.path.exists(p):
                    print(f"Tìm thấy '{file_name}' tại đường dẫn WSL: {p}")
                    return os.path.abspath(p)

    # 3. Leo ngược cây thư mục tìm kiếm thư mục có chứa 'Du_AN' hoặc 'data'
    curr = os.path.abspath(os.getcwd())
    for _ in range(6):
        if os.path.exists(curr):
            try:
                items = os.listdir(curr)
                for item in items:
                    if item.lower() in ['du_an', 'data']:
                        target_dir = os.path.join(curr, item)
                        for root, dirs, files in os.walk(target_dir):
                            depth = root.replace(target_dir, "").count(os.sep)
                            if depth > 2:
                                dirs.clear()  # Cắt tỉa nhánh tìm kiếm để tối ưu tốc độ
                                continue
                            for f in files:
                                if f.lower() == file_name.lower():
                                    p = os.path.join(root, f)
                                    print(f"Tìm thấy '{file_name}' (quét dự án) tại: {p}")
                                    return os.path.abspath(p)
            except Exception:
                pass
        parent = os.path.dirname(curr)
        if parent == curr:
            break
        curr = parent
                    
    # Mặc định tương đối
    return os.path.join("..", "data", "raw", file_name)

playstore_path = find_data_file("googleplaystore.csv")
reviews_path = find_data_file("googleplaystore_user_reviews.csv")

# Sử dụng đường dẫn tuyệt đối (abspath) trước khi trích xuất thư mục cha để tránh lỗi đường dẫn tương đối trống
abs_playstore = os.path.abspath(playstore_path)
raw_dir = os.path.dirname(abs_playstore)
data_dir = os.path.dirname(raw_dir)
project_dir = os.path.dirname(data_dir)

processed_dir = os.path.join(data_dir, "processed")
sql_dir = os.path.join(project_dir, "sql")

# Tạo thư mục processed nếu chưa tồn tại để tránh lỗi pandas to_csv
os.makedirs(processed_dir, exist_ok=True)

print("Đường dẫn Raw Apps:", playstore_path)
print("Đường dẫn Raw Reviews:", reviews_path)
print("Thư mục Processed:", processed_dir)

# Đọc dữ liệu raw từ đĩa
df_apps_raw = pd.read_csv(playstore_path)
df_reviews_raw = pd.read_csv(reviews_path)

# Chuẩn hóa tên cột để chạy trên các lệnh SQL chuẩn
df_apps_raw.columns = [c.replace(' ', '_') for c in df_apps_raw.columns]
df_reviews_raw.columns = [c.replace(' ', '_') for c in df_reviews_raw.columns]

# Ghi dữ liệu vào các bảng staging trong database giả lập
df_apps_raw.to_sql('googleplaystore', sim_conn, index=False, if_exists='replace')
df_reviews_raw.to_sql('googleplaystore_user_reviews', sim_conn, index=False, if_exists='replace')

print("--- KIỂM TRA SỐ DÒNG SAU KHI IMPORT (COUNT(*)) ---")
query_count = """
SELECT 'googleplaystore' AS table_name, COUNT(*) AS row_count FROM googleplaystore
UNION ALL
SELECT 'googleplaystore_user_reviews' AS table_name, COUNT(*) AS row_count FROM googleplaystore_user_reviews;
"""
print(pd.read_sql(query_count, sim_conn))


=== CHẨN ĐOÁN HỆ THỐNG & ĐƯỜNG DẪN ===
Thư mục đang chạy (getcwd): C:\HOC_HANH\Mon dang hoc\AI21301_DU_AN_1
Tìm thấy 'googleplaystore.csv' tại đường dẫn tương đối: Du_AN/data/raw\googleplaystore.csv
Tìm thấy 'googleplaystore_user_reviews.csv' tại đường dẫn tương đối: Du_AN/data/raw\googleplaystore_user_reviews.csv
Đường dẫn Raw Apps: C:\HOC_HANH\Mon dang hoc\AI21301_DU_AN_1\Du_AN\data\raw\googleplaystore.csv
Đường dẫn Raw Reviews: C:\HOC_HANH\Mon dang hoc\AI21301_DU_AN_1\Du_AN\data\raw\googleplaystore_user_reviews.csv
Thư mục Processed: C:\HOC_HANH\Mon dang hoc\AI21301_DU_AN_1\Du_AN\data\processed
--- KIỂM TRA SỐ DÒNG SAU KHI IMPORT (COUNT(*)) ---
                     table_name  row_count
0               googleplaystore      10841
1  googleplaystore_user_reviews      64295


**Nhận xét:**
- Bảng chính `googleplaystore` nạp thành công **10,841 dòng**.
- Bảng phụ `googleplaystore_user_reviews` nạp thành công **64,295 dòng**.


# V. Tối ưu hóa Database (Tạo INDEX)
Để tăng tốc truy vấn khi kết hợp (JOIN) và gom nhóm (GROUP BY) bảng dữ liệu lớn, ta thực hiện tạo các INDEX trên trường khóa liên kết `App`.


In [1]:
cursor = sim_conn.cursor()
cursor.execute("CREATE INDEX IF NOT EXISTS idx_apps_app ON googleplaystore(App);")
cursor.execute("CREATE INDEX IF NOT EXISTS idx_reviews_app ON googleplaystore_user_reviews(App);")
sim_conn.commit()

print("Đã khởi tạo thành công các INDEX trên cột 'App' ở cả hai bảng staging.")


Đã khởi tạo thành công các INDEX trên cột 'App' ở cả hai bảng staging.


# VI. Tích hợp dữ liệu và Tránh Nhân bản dòng (Summary & Aggregation View)
> [!IMPORTANT]
> **Hiện tượng nhân bản dòng khi JOIN 1 - N:**
> Bảng chính `googleplaystore` (mỗi app có 1 dòng) và bảng phụ `googleplaystore_user_reviews` (mỗi app có nhiều dòng reviews) có mối quan hệ 1-N. Nếu ta `LEFT JOIN` trực tiếp, số dòng của bảng chính sẽ bị nhân lên nhiều lần (lên tới 64k+ dòng), làm méo mó nghiêm trọng phân phối của Rating và Installs phục vụ Machine Learning.
> 
> **Giải pháp:**
> Xây dựng **Summary View** `vw_reviews_summary` để thực hiện gom nhóm (`GROUP BY App`) và tổng hợp dữ liệu nhận xét của người dùng trước khi JOIN. Mỗi ứng dụng sẽ chỉ còn đúng 1 dòng đại diện.

### Câu lệnh SQL tạo Aggregation View (`04_aggregation.sql`)
```sql
CREATE VIEW vw_reviews_summary AS
SELECT 
    App,
    COUNT(*) AS total_reviews_count,
    ROUND(AVG(CAST(Sentiment_Polarity AS NUMERIC)), 4) AS avg_sentiment_polarity,
    ROUND(AVG(CAST(Sentiment_Subjectivity AS NUMERIC)), 4) AS avg_sentiment_subjectivity
FROM googleplaystore_user_reviews
WHERE App IS NOT NULL AND Translated_Review IS NOT NULL
GROUP BY App;
```


In [1]:
query_summary = """
CREATE VIEW IF NOT EXISTS vw_reviews_summary AS
SELECT 
    App,
    COUNT(*) AS total_reviews_count,
    ROUND(AVG(CAST(Sentiment_Polarity AS NUMERIC)), 4) AS avg_sentiment_polarity,
    ROUND(AVG(CAST(Sentiment_Subjectivity AS NUMERIC)), 4) AS avg_sentiment_subjectivity
FROM googleplaystore_user_reviews
WHERE App IS NOT NULL AND Translated_Review IS NOT NULL
GROUP BY App;
"""
cursor.execute(query_summary)
sim_conn.commit()

df_summary = pd.read_sql("SELECT * FROM vw_reviews_summary LIMIT 5;", sim_conn)
print("--- 5 DÒNG TỔNG HỢP CẢM XÚC CỦA MẪU ỨNG DỤNG ---")
print(df_summary)


--- 5 DÒNG TỔNG HỢP CẢM XÚC CỦA MẪU ỨNG DỤNG ---
                                App  ...  avg_sentiment_subjectivity
0             10 Best Foods for You  ...                      0.4955
1  104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室  ...                      0.5455
2                              11st  ...                      0.4553
3        1800 Contacts - Lens Store  ...                      0.5911
4   1LINE – One Line with One Touch  ...                      0.5573

[5 rows x 4 columns]


# VII. Khởi tạo Flat View hoàn chỉnh (vw_googleplaystore_flat)
Tiến hành kết nối bảng chính với Summary View để thu được bảng phẳng cuối cùng, phục vụ tiền xử lý và huấn luyện AI.

### Câu lệnh SQL tạo Flat View (`03_views.sql`)
```sql
CREATE VIEW vw_googleplaystore_flat AS
SELECT 
    a.App,
    a.Category,
    a.Rating,
    a.Reviews,
    a.Size,
    a.Installs,
    a.Type,
    a.Price,
    a.Content_Rating,
    a.Genres,
    a.Last_Updated,
    a.Current_Ver,
    a.Android_Ver,
    COALESCE(r.total_reviews_count, 0) AS total_user_reviews,
    COALESCE(r.avg_sentiment_polarity, 0.0) AS avg_sentiment_polarity,
    COALESCE(r.avg_sentiment_subjectivity, 0.0) AS avg_sentiment_subjectivity
FROM googleplaystore a
LEFT JOIN vw_reviews_summary r ON a.App = r.App;
```


In [1]:
query_flat = """
CREATE VIEW IF NOT EXISTS vw_googleplaystore_flat AS
SELECT 
    a.App,
    a.Category,
    a.Rating,
    a.Reviews,
    a.Size,
    a.Installs,
    a.Type,
    a.Price,
    a.Content_Rating,
    a.Genres,
    a.Last_Updated,
    a.Current_Ver,
    a.Android_Ver,
    COALESCE(r.total_reviews_count, 0) AS total_user_reviews,
    COALESCE(r.avg_sentiment_polarity, 0.0) AS avg_sentiment_polarity,
    COALESCE(r.avg_sentiment_subjectivity, 0.0) AS avg_sentiment_subjectivity
FROM googleplaystore a
LEFT JOIN vw_reviews_summary r ON a.App = r.App;
"""
cursor.execute(query_flat)
sim_conn.commit()

# Đọc thử Flat View từ Database
df_flat = pd.read_sql("SELECT * FROM vw_googleplaystore_flat;", sim_conn)
print(f"Số dòng bảng chính ban đầu: {df_apps_raw.shape[0]:,}")
print(f"Số dòng bảng phẳng sau khi JOIN: {df_flat.shape[0]:,}")
print(f"Số cột của bảng phẳng: {df_flat.shape[1]}")


Số dòng bảng chính ban đầu: 10,841
Số dòng bảng phẳng sau khi JOIN: 10,841
Số cột của bảng phẳng: 16


**Nhận xét:**
- Số lượng dòng sau khi JOIN **hoàn toàn trùng khớp (10,841 dòng)**. Quá trình tích hợp dữ liệu diễn ra hoàn hảo mà không làm phát sinh dòng trùng lặp hay làm mất dữ liệu gốc.


# VIII. Kiểm tra Hiệu năng kết nối & Truy vấn (Performance Verification)
Sử dụng câu lệnh Python đo đạc thời gian truy xuất dữ liệu từ Database để đảm bảo tốc độ phản hồi tối ưu.


In [1]:
import time

start_time = time.time()
df_perf = pd.read_sql("SELECT * FROM vw_googleplaystore_flat;", sim_conn)
elapsed = time.time() - start_time

print(f"Thời gian truy vấn đọc toàn bộ bảng phẳng Flat View (10,841 dòng): {elapsed:.4f} giây")


Thời gian truy vấn đọc toàn bộ bảng phẳng Flat View (10,841 dòng): 0.0801 giây


# IX. Kết luận & Giải đáp 4 câu hỏi cốt lõi

### 1. Dữ liệu đã được đưa vào PostgreSQL đúng chưa?
- **Trả lời:** Đúng. Toàn bộ 10,841 dòng của bảng Apps và 64,295 dòng của bảng Reviews đã được nạp và đối chiếu số lượng thành công thông qua lệnh `COUNT(*)`.

### 2. Làm thế nào để tích hợp dữ liệu từ nhiều bảng?
- **Trả lời:** Ta đã sử dụng kỹ thuật gom nhóm dữ liệu (Aggregation) theo `App` ở bảng phụ Reviews nhằm đưa mối quan hệ 1-N về quan hệ 1-1, sau đó tiến hành `LEFT JOIN` với bảng chính để không bị nhân bản dòng.

### 3. Dữ liệu nào sẽ được dùng cho Machine Learning?
- **Trả lời:** Flat View `vw_googleplaystore_flat` chứa đầy đủ 13 trường thông tin metadata của ứng dụng cùng với 3 trường thông tin nhận xét tổng hợp của người dùng (tổng số review, độ phân cực cảm xúc trung bình, độ chủ quan trung bình).

### 4. Pipeline dữ liệu đã sẵn sàng cho Notebook 03 chưa?
- **Trả lời:** Đã sẵn sàng. Kể từ Notebook 03 trở đi, toàn bộ quá trình đọc dữ liệu sẽ được thực thi trực tiếp bằng cách gọi truy vấn SQL từ Database Flat View, đảm bảo tính đồng bộ dữ liệu cao nhất.
